khook/x86/hook.c

```diff
@@ -1,22 +1,112 @@
 #include "../internal.h"
 
-#define kernel_write_enter() asm volatile (	\
+#define ASM_CLI asm volatile ("cli\n\t" ::: )
+#define ASM_STI asm volatile ("sti\n\t" ::: )
+
+#ifndef USE_CR0_FAM
+# ifndef USE_PTE_FAM
+#  error "At least one FAM *must* be used..."
+# endif
+#endif
+
+#ifdef USE_CR0_FAM
+# define kernel_write_enter() asm volatile (	\
 	"cli\n\t"				\
 	"mov %%cr0, %%rax\n\t"			\
 	"and $0xfffffffffffeffff, %%rax\n\t"	\
 	"mov %%rax, %%cr0\n\t"			\
 	::: "%rax" )
 
-#define kernel_write_leave() asm volatile (	\
+# define kernel_write_leave() asm volatile (	\
 	"mov %%cr0, %%rax\n\t"			\
 	"or $0x0000000000010000, %%rax\n\t"	\
 	"mov %%rax, %%cr0\n\t"			\
 	"sti\n\t"				\
 	::: "%rax" )
-
+#endif
+#ifdef USE_PTE_FAM
+/* Because of syncronization mechanism
+ * this set_rw/set_ro can be used only pairy
+ * and should be used in small time window
+ **/
+static int
+	is_first_rw = 0x00,
+	is_last_rw = 0x00;
+static volatile int lock_pte = 0x00;
+/*|SPIN FAMILY|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//**/
```

Comment:
Please, do not use `/*|*//*|....|*/` comments, there is not reason to have them.

Reply:
But, in *.c files we should use `/* */` and in the *.h files we should use `//`, doesn't we?

Reply:
I mean no comments like that at all